packages/gatsby/src/internal-plugins/query-runner/gatsby-node.js

```diff
@@ -1,31 +1,42 @@
 const fs = require(`fs`)
 const path = require(`path`)
-
 const { watchComponent } = require(`./query-watcher`)
 
-let pageComponents = {}
-exports.onCreatePage = ({ page, store, boundActionCreators }) => {
-  const component = page.component
-  if (!pageComponents[component]) {
-    // We haven't seen this component before so we:
-    // - Ensure it has a JSON file.
-    // - Add it to Redux
-    // - Watch the component to detect query changes
-    const pathToJSONFile = path.join(
+let components = {}
+
+const handlePageOrLayout = store => pageOrLayout => {
+  // - ensure corresponding page or layout has json files.
+  // - get corresponding component
+  // - watch component
+  // - mark component
+  const writeJsonFile = ({ jsonName }) => {
+    const dest = path.join(
       store.getState().program.directory,
       `.cache`,
       `json`,
-      page.jsonName
+      jsonName
     )
-    if (!fs.existsSync(pathToJSONFile)) {
-      fs.writeFile(pathToJSONFile, `{}`, () => {})
+    if (!fs.existsSync(dest)) {
+      fs.writeFile(dest, `{}`, () => {})
     }
-    boundActionCreators.createPageComponent(component)
+  }
+
+  writeJsonFile(pageOrLayout)
+
+  const component = store.getState().components[pageOrLayout.componentPath]
 
-    // Make sure we're watching this component.
-    watchComponent(component)
+  if (components[component.componentPath]) {
+    return
   }
 
-  // Mark we've seen this page component.
-  pageComponents[component] = component
+  watchComponent(component.componentPath)
+  components[component.componentPath] = component.componentPath
+}
+
+exports.onCreatePage = ({ page, store, boundActionCreators }) => {
+  handlePageOrLayout(store)(page)
```

Comment:
None here. I just prefer to compose in that was so I don't refactor later we it is useful. Happy to rewrite if everyone pefers multi args